ppci/lang/c/semantics.py

```diff
@@ -692,21 +693,69 @@ def on_binop(self, lhs, op, rhs, location):
             result_typ = lhs.typ
             if not lhs.lvalue:
                 self.error("Expected lvalue", lhs.location)
+
+            if op[0] in ["+", "-"] and isinstance(lhs.typ, types.IndexableType):
+                self.ensure_integer(rhs.typ, location)
+                lhs.typ = self.ensure_no_void_ptr(lhs.typ, location)
```

Comment:
Note, I implemented this slightly different by introducing a cast instruction from `void*` to `char*`. In spirit of rust, I think immutable values are a good thing, that why I left the original `lhs` as is and added a cast.